### PR TITLE
dlopen -> dlclose -> dlopen bug

### PR DESCRIPTION
**Pull request by wowlocal**
Add FIRCLSBinaryImageRecordingTests
Make FIRCLSContextInitialize to initialized once per unit tests run
Replace FIRCLSTempMockFileManager with FIRCLSFileManager

### CODE REVIEW

**Comment by samedson** [comment]
Thank you for contributing! I'll take a pass at reviewing this tomorrow.

Could you describe the bug this is fixing a little more?

**Comment by wowlocal** [comment]
> Thank you for contributing! I'll take a pass at reviewing this tomorrow.
> 
> Could you describe the bug this is fixing a little more?

I tried to describe the problem in the comments to the code. Take a look into FIRCLSBinaryImageRecordingTests.m
The first thing that comes to mind is to change search algorithm inside FIRCLSBinaryImageSafeFindImageForAddress to the reverse order or to make complete removal of a node binaryImage.nodes list. But i feel like there is a better way.

File: Crashlytics/UnitTests/Mocks/FIRCLSTempMockFileManager.h [hunk 1]
```diff
@@ -1,48 +0,0 @@
...
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-#import "Crashlytics/Crashlytics/Models/FIRCLSFileManager.h"
-
-#import <XCTest/XCTest.h>
-
-@interface FIRCLSTempMockFileManager : FIRCLSFileManager
```
Comment (samedson) [inline Crashlytics/UnitTests/Mocks/FIRCLSTempMockFileManager.h:19]: I'm curious about the reason behind removing the TempMockFileManager. Were you happening to see flakes in the tests before this refactor?

**Comment by wowlocal** [comment]
> When you run tests, do you also see this crash?
> 
> 
> 
> Screen Shot 2022-07-15 at 1 46 44 PM
> 
> 

Yep, I guess it has something to do with ARM on macOS. Tests on Rosetta seem to work properly

**Comment by wowlocal** [comment]
> So if I'm summarizing this PR:
> 
>  - It sounds like there's a bug where if you dlopen, close, and open again, Crashlytics can run into issues. 
> 
>  - The tests you added illustrate this bug
> 
>  - There is still work left to figure out how to fix the bug
> 
> 
> 
> Does that match your understanding?

You are 100% right! :) Currently I'm limited in time but I will dive deeper into this bug later

**Comment by samedson** [comment]
@wowlocal we're going through and closing out old threads. I'm going to go ahead and close this. If you find the time or want to talk more about it feel free to reopen.

**Pull request closed without merging**